nautobot/dcim/models/devices.py

```diff
@@ -1375,3 +1386,329 @@ def clean(self):
+            return None
+        return self.parent_module_bay.device
+
+    def clean(self):
+        super().clean()
+
+        # Validate that the Module is associated with a Location or a ModuleBay
+        if self.parent_module_bay is None and self.location is None:
+            raise ValidationError("One of location or parent_module_bay must be set")
+
+        # Validate location
+        if self.location is not None:
+            if self.parent_module_bay is not None:
+                raise ValidationError("Only one of location or parent_module_bay must be set")
+
+            if ContentType.objects.get_for_model(self) not in self.location.location_type.content_types.all():
+                raise ValidationError(
+                    {"location": f'Modules may not associate to locations of type "{self.location.location_type}".'}
+                )
+
+    def save(self, *args, **kwargs):
+        is_new = not self.present_in_database
+
+        if self.serial == "":
+            self.serial = None
+        if self.asset_tag == "":
+            self.asset_tag = None
+
+        super().save(*args, **kwargs)
+
+        # If this is a new Module, instantiate all related components per the ModuleType definition
+        if is_new:
+            self.create_components()
+
+    def create_components(self):
+        """Create module components from the module type definition."""
+        # The order of these is significant as
+        # - PowerOutlet depends on PowerPort
+        # - FrontPort depends on RearPort
+        component_models = [
+            (ConsolePort, self.module_type.console_port_templates.all()),
+            (ConsoleServerPort, self.module_type.console_server_port_templates.all()),
+            (PowerPort, self.module_type.power_port_templates.all()),
+            (PowerOutlet, self.module_type.power_outlet_templates.all()),
+            (Interface, self.module_type.interface_templates.all()),
+            (RearPort, self.module_type.rear_port_templates.all()),
+            (FrontPort, self.module_type.front_port_templates.all()),
+            (ModuleBay, self.module_type.module_bay_templates.all()),
+        ]
+        instantiated_components = []
+        for model, templates in component_models:
+            model.objects.bulk_create([x.instantiate(device=None, module=self) for x in templates])
+        return instantiated_components
+
+    def get_cables(self, pk_list=False):
+        """
+        Return a QuerySet or PK list matching all Cables connected to a component of this Module.
+        """
+        from .cables import Cable
+
+        cable_pks = []
+        for component_model in [
+            ConsolePort,
+            ConsoleServerPort,
+            PowerPort,
+            PowerOutlet,
+            Interface,
+            FrontPort,
+            RearPort,
+        ]:
+            cable_pks += component_model.objects.filter(module=self, cable__isnull=False).values_list(
+                "cable", flat=True
```

Comment:
Should that be `"cable_id"` instead?

Reply:
This was copy/pasted from Device and it works there:

```py
>>> Device.objects.first().get_cables()
<RestrictedQuerySet [<Cable: #408b097b-4011-40eb-b448-62e5ae9183f2>, <Cable: #2ea6dbac-9936-4207-9d30-61c5a1b3c8a2>]>
```